Add ImageSharp benchmark

kunalspathak opened the pull request:
This is a snapshot taken from https://github.com/SixLabors/ImageSharp/commit/413c90c23d448480906f82bf742feeecfb7f4eb1. I deleted `System.Drawing*` related benchmarks and most of the other 3rd party baseline benchmarks to keep things simple. I also copied few `TestUtilities` related code and the input images.
I ran it on Windows/x64 and Windows/arm64 and they seem to be running fine. It needs some more refining to get 100% pass rate.

Relevant discussion: https://github.com/SixLabors/ImageSharp/discussions/1795#discussioncomment-1635798

FYI - @antonfirsov , @JimBobSquarePants

JimBobSquarePants commented:
Will have a look ASAP. A little concerned about how to maintain this. 

Ideally I would say that there would be a simple set of benchmarks that use a reference to ImageSharp via NuGet. That way anyone can PR to update the references when required.

danmoseley commented:
if you go ahead with this, there probably should be an entry in THIRD_PARTY_NOTICES.TXT

I agree with @JimBobSquarePants isn't a NuGet reference to a fixed package version sufficient? There is a lot of code here that requires building, etc.

kunalspathak commented:
> that use a reference to ImageSharp via NuGet

That's what I have currently. It reference to the ImageSharp version [1.0.2 via nuget](https://github.com/kunalspathak/performance/blob/9df3ef61762c363b7c063a244f930dff397c85e9/src/benchmarks/real-world/ImageSharp/ImageSharp.Benchmarks.csproj#L23-L24). The reason there are so many files are because I copied every single benchmark file present in the original repo modulo the other baseline benchmarks. Perhaps, if there is a subset of benchmarks that you think can represent a good coverage of ImageSharp, feel free to point out and I am happy to trim it to that subset.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/General/PixelConversion/PixelConversion_Rgba32_To_Bgra32.cs:
```diff
@@ -0,0 +1,357 @@
+{
+    // [MonoJob]
+    // [RyuJitX64Job]
+    public class PixelConversion_Rgba32_To_Bgra32
```
Same.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/General/StructCasting.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace SixLabors.ImageSharp.Benchmarks.General
+{
+    public class StructCasting
```
Doesn't use the library.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/General/Vector4Constants.cs:
```diff
@@ -0,0 +1,62 @@
+    /// Has it any effect on performance to store SIMD constants as static readonly fields? Is it OK to always inline them?
+    /// Spoiler: the difference seems to be statistically insignificant!
+    /// </summary>
+    public class Vector4Constants
```
Doesn't use the library.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/General/Vectorization/BitwiseOrUint32.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace SixLabors.ImageSharp.Benchmarks.General.Vectorization
+{
+    public class BitwiseOrUInt32
```
Doesn't use the library.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/General/Vectorization/DivFloat.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace SixLabors.ImageSharp.Benchmarks.General.Vectorization
+{
+    public class DivFloat
```
I will stop commenting on these. `General/Vectorization` or better, the whole `General` folder should be dropped.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Images/Input/Jpg/baseline/JpegSnoopReports/Calliphora.jpg.txt:
```diff
@@ -0,0 +1,331 @@
+
+JPEGsnoop 1.8.0 by Calvin Hass
```
These files are not used by benchmarks, it' just for us to know what are the test images about.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Images/Input/Tiff/Benchmarks/gen_big.ps1:
```diff
@@ -0,0 +1,14 @@
+$Gm_Exe = "C:\Program Files\ImageMagick-7.0.10-Q16-HDRI\magick.exe"
```
I'm not sure about tiff benchmarks, but I doubt you want to run benchmarks on CI which need files generated by ImageMagick :)

antonfirsov commented on src/benchmarks/real-world/ImageSharp/LoadResizeSave/LoadResizeSaveStressRunner.cs:
```diff
@@ -0,0 +1,157 @@
+        Any = Baseline | Progressive
+    }
+
+    public class LoadResizeSaveStressRunner
```
Remove everything that uses this, and/or the whole `LoadResizeSave` folder.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Processing/DetectEdges.cs:
```diff
@@ -0,0 +1,48 @@
+        [Benchmark(Description = "ImageSharp DetectEdges")]
+        public void ImageProcessorCoreDetectEdges()
+        {
+            this.image.Mutate(x => x.DetectEdges(KnownEdgeDetectorKernels.Kayyali));
```
Only keep one these calls, doesn't really matter which one.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Processing/Diffuse.cs:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+// #### 20th February 2020 ####
+//
+// BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
+// Intel Core i7-8650U CPU 1.90GHz(Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
+// .NET Core SDK = 3.1.101
+//
+// [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
+//  Job-OJKYBT : .NET Framework 4.8 (4.8.4121.0), X64 RyuJIT
+//  Job-RZWLFP : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), X64 RyuJIT
+//  Job-NUYUQV : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
+//
+// IterationCount=3  LaunchCount=1  WarmupCount=3
+//
+// |    Method |       Runtime |      Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
+// |---------- |-------------- |----------:|----------:|----------:|------:|------:|------:|----------:|
+// | DoDiffuse |    .NET 4.7.2 | 30.535 ms | 19.217 ms | 1.0534 ms |     - |     - |     - |  26.25 KB |
+// |  DoDither |    .NET 4.7.2 | 14.174 ms |  1.625 ms | 0.0891 ms |     - |     - |     - |  31.38 KB |
+// | DoDiffuse | .NET Core 2.1 | 15.984 ms |  3.686 ms | 0.2020 ms |     - |     - |     - |  25.98 KB |
+// |  DoDither | .NET Core 2.1 |  8.646 ms |  1.635 ms | 0.0896 ms |     - |     - |     - |  28.99 KB |
+// | DoDiffuse | .NET Core 3.1 | 16.235 ms |  9.612 ms | 0.5269 ms |     - |     - |     - |  25.96 KB |
+// |  DoDither | .NET Core 3.1 |  8.429 ms |  1.270 ms | 0.0696 ms |     - |     - |     - |  31.61 KB |
```
You probably want to remove these comments from the benchmarks, if not too much chore?

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Processing/Skew.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace SixLabors.ImageSharp.Benchmarks.Processing
+{
+    public class Skew
```
Drop this, internally it's the same as Rotate.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Program.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <param name="args">
+        /// The arguments to pass to the program.
+        /// </param>
+        public static void Main(string[] args) => BenchmarkSwitcher
```
Do you need our `Program.cs`?

kunalspathak replied:
For now, yes. That's how we have our other real-world benchmarks setup.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/TestUtilities/TestEnvironment.cs:
```diff
@@ -0,0 +1,282 @@
+
+namespace SixLabors.ImageSharp.Tests
+{
+    public static partial class TestEnvironment
```
You may want to drop the unused bits from here. (The original file is our test infra god-object.)

antonfirsov commented on src/benchmarks/real-world/ImageSharp/TestUtilities/TestImages.cs:
```diff
@@ -0,0 +1,682 @@
+    /// Class that contains all the relative test image paths in the TestImages/Formats directory.
+    /// Use with <see cref="WithFileAttribute"/>, <see cref="WithFileCollectionAttribute"/>.
+    /// </summary>
+    public static class TestImages
```
I would review this class and only keep the images which are actually used, deleting the rest of the image files. (I know it's a lot of chore work ...)

kunalspathak commented on src/benchmarks/real-world/ImageSharp/Codecs/Jpeg/DecodeJpeg_ImageSpecific.cs:
```diff
@@ -0,0 +1,69 @@
+    /// Image-specific Jpeg benchmarks
+    /// </summary>
+    
+    public class DecodeJpeg_ImageSpecific
```
Are you suggesting to use `DecodeJpeg.cs` file as it is?

br3aker replied:
I would say that `DecodeJpeg.cs` is the best bet atm but it has some variance in results. 3 out of 4 images there are stable in terms of execution time but the first one has a stable delta of 0.5ms between different benchmark runs. I assume it is caused by the GC but I'm not sure.

kunalspathak commented:
> if you go ahead with this, there probably should be an entry in THIRD_PARTY_NOTICES.TXT

Done.

kunalspathak commented:
> If you manage to get rid of the noisy stuff, I can do another review.

@antonfirsov - Thanks for taking time and providing feedback. I have deleted all the benchmarks that you have suggested. Do you mind doing another round of review?

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/DecodeFilteredPng.cs:
```diff
@@ -0,0 +1,60 @@
+            this.filter4 = File.ReadAllBytes(TestImageFullPath(TestImages.Png.Filter4));
+        }
+
+        [Benchmark(Baseline = true, Description = "None-filtered PNG file")]
```
```suggestion
        [Benchmark(Description = "None-filtered PNG file")]
```

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodeBmp.cs:
```diff
@@ -0,0 +1,42 @@
+        [Benchmark(Description = "ImageSharp Bmp")]
+        public void BmpImageSharp()
+        {
+            using var memoryStream = new MemoryStream();
```
This stream should be a member, initialized in `[GlobalSetup]`, and `Seek`-ed to `0` before the `SaveAsBmp` call. We don't want the variance and the GC pressure stream allocation is causing here.

kunalspathak replied:
But since `BmpImageSharp()` will be called repeatedly, won't it continue saving and growing the memory stream and increase memory pressure?

antonfirsov replied:
If you seek to the beginning before saving, it only grows if you exceed the output size of the previous iteration, which is not the case, since you are saving a file of the same image deterministically. Or am I missing something?

antonfirsov replied:
And that's the good thing. Are we are missing each others points?

This is what I mean:

```C#
private MemoryStream memoryStream;

[GlobalSetup]
public void Setup()
{
    this.memoryStream = new MemoryStream();
}

[Benchmark]
public void DoBenchmark()
{
    this.memoryStream.Seek(0, SeekOrigin.Begin); // avoid growing the buffer if we write the same amount of data over and over

   this.image.SaveAsBmp(this.memoryStream);
}
```

kunalspathak replied:
Ah, I thought you wanted to do `this.memoryStream.Seek(0)` inside `GlobalSetup`. Will fix it. Thanks!

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodeGif.cs:
```diff
@@ -0,0 +1,62 @@
+            this.bmpCore.Dispose();
+        }
+
+        [Benchmark(Baseline = true, Description = "System.Drawing Gif")]
+        public void GifSystemDrawing()
+        {
+            using var memoryStream = new MemoryStream();
+        }
```
```suggestion
```

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodeGif.cs:
```diff
@@ -0,0 +1,62 @@
+        [Benchmark(Description = "ImageSharp Gif")]
+        public void GifImageSharp()
+        {
+            using var memoryStream = new MemoryStream();
```
Same `memoryStream` issue as with `EncodeBmp`.

/cc @JimBobSquarePants we should also cleanup these.

kunalspathak replied:
See my comments above.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodeIndexedPng.cs:
```diff
@@ -0,0 +1,91 @@
+        [Benchmark(Description = "ImageSharp Octree Png")]
+        public void PngCoreOctree()
+        {
+            using var memoryStream = new MemoryStream();
```
All the MemoryStreams 😱

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodePng.cs:
```diff
@@ -0,0 +1,50 @@
+        [Benchmark(Description = "ImageSharp Png")]
+        public void PngCore()
+        {
+            using var memoryStream = new MemoryStream();
```
!

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/EncodeTga.cs:
```diff
@@ -0,0 +1,44 @@
+        [Benchmark(Description = "ImageSharp Tga")]
+        public void ImageSharpTga()
+        {
+            using var memoryStream = new MemoryStream();
```
!

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Codecs/Jpeg/DecodeJpeg_ImageSpecific.cs:
```diff
@@ -0,0 +1,56 @@
+    /// </summary>
+    
+    public class DecodeJpeg_ImageSpecific
+    {
```
@kunalspathak any chance you can update this as discussed above?

kunalspathak replied:
Done.

antonfirsov replied:
It's not pushed yet?

antonfirsov reviewed (commented):
If you pull in the new Jpeg benchmarks, add [Resize](https://github.com/SixLabors/ImageSharp/blob/master/tests/ImageSharp.Benchmarks/Processing/Resize.cs) benchmark, fix the `memoryStream` issues this should be mostly good to go.

Any insights whether the variance of execution times is within acceptable range for these benchmarks?

antonfirsov commented:
Regarding the resize benchmark:

We don't need the [`maxDegreeOfParallelism` stuff](https://github.com/SixLabors/ImageSharp/blob/8fbc68cc4cac9f43951322efead1c7f50d709f27/tests/ImageSharp.Benchmarks/Processing/Resize.cs#L88) it's outdated stuff, resize implementation is not parallel anymore.

The subclasses exercising different scenarios are valuable, except `Resize_Bicubic_Rgba32_CompareWorkBufferSizes`, which can be trimmed.

danmoseley commented:
@kunalspathak are you still planning to continue with this?

kunalspathak commented:
> @kunalspathak are you still planning to continue with this?

Yes. I have been busy with Arm64 investigation work. Will get to this soon. I don't think much work is needed here.

kunalspathak commented:
Ping to myself.

kunalspathak commented:
@antonfirsov , @JimBobSquarePants - I will resume working on this project now. Wondering what should be the next steps? Should I update the ImageSharp code to more latest or should we get this in as-is (after addressing the review comments) and then do another PR to update the code base to the latest?

JimBobSquarePants commented:
If you are going to continue the snapshot approach I would recommend updating to the [latest V2.x version](https://github.com/SixLabors/ImageSharp/tree/release/2.1.x) before continuing. V1 numbers won't be beneficial and I can imagine you don't want to go through the PR process multiple times.

However, for ease of maintenance I would suggest that the benchmarks are performed using NuGet packages.

kunalspathak commented:
> If you are going to continue the snapshot approach I would recommend updating to the [latest V2.x version](https://github.com/SixLabors/ImageSharp/tree/release/2.1.x) before continuing. V1 numbers won't be beneficial and I can imagine you don't want to go through the PR process multiple times.
> 

I agree. I will update the version in https://github.com/dotnet/performance/pull/2149/files#diff-99843982e8a5e1acef45cd069c3e3230ec1a014e7fe01aea4c05f2339b40afbc.

> However, for ease of maintenance I would suggest that the benchmarks are performed using NuGet packages.

As part of this PR, ImageSharp will be consumed via nuget packages, if that's what you meant by "benchmarks are performed using NuGet packages.". My question was more around if I should take the latest benchmarks source code from ImageSharp repo or not.

kunalspathak commented:
> My question was more around if I should take the latest benchmarks source code from ImageSharp repo or not.

I prefer taking the latest benchmark code and if we go this route, may be it will be helpful to point which files/folders to take vs. omit as was done previously as part of the review comments.

antonfirsov commented:
> I prefer taking the latest benchmark code

I recommend to continue the work based on the current state. There was a lot of sanitation work in this PR which is better not to redo.

What's still missing (also see previous comments like https://github.com/dotnet/performance/pull/2149#pullrequestreview-823293095):
- Move `MemoryStream` initialization to setup phase
- Add Resize benchmark
- Replace `DecodeJpeg_ImageSpecific` with the new Jpeg benchmarks landed just a day after opening this PR -- see https://github.com/dotnet/performance/pull/2149/files#r755372378

kunalspathak commented:
> There was a lot of sanitation work in this PR which is better not to redo.

Agree. But I guess we might have to do similar work every time we would update the benchmarks unless you have some better way/docs to tell what things to copy over and what not. For example, my plan after merging this PR is to do another sync of benchmarks so we are at least up-to-date on benchmarks side when we start measuring the perf of ImageSharp.

JimBobSquarePants commented:
>For example, my plan after merging this PR is to do another sync of benchmarks so we are at least up-to-date on benchmarks side when we start measuring the perf of ImageSharp.

It's probably better to maintain a sensible subset of benchmarks that don't have to be resynced.

- Decode
- Encode
- Resize

Our processing pipelines reuse a lot of underlying code so any issues that affect other areas should become apparent. If we spot anything in the future missed here we can add additional benchmarks when required. Thoughts @antonfirsov

antonfirsov commented:
> It's probably better to maintain a sensible subset of benchmarks that don't have to be resynced.

Agreed. The benchmark code in the ImageSharp repo is of varying quality, made with different priorities than the ones of this repo and a sandboxing approach. This is not something you want to continuously sync back.

kunalspathak commented:
> It's probably better to maintain a sensible subset of benchmarks that don't have to be resynced.
> 
> * Decode
> * Encode
> * Resize

I see `Decode` and `Encode` benchmarks inside `Codecs` folder. Could you confirm which ones are `Resize` related and what other benchmarks I should remove in this PR?

antonfirsov commented:
- Resize benchmarks can be based on [this code](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Benchmarks/Processing/Resize.cs), but have to be heavily sanitized, see https://github.com/dotnet/performance/pull/2149#issuecomment-986080701.
- The jpeg decoder benchmark `DecodeJpeg_ImageSpecific` should be replaced with `DecodeJpeg` from https://github.com/SixLabors/ImageSharp/pull/1844
- I would consider removing the benchmarks under the folder `Color`, it's an unfinished niche feature

kunalspathak commented:
> * Resize benchmarks can be based on [this code](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Benchmarks/Processing/Resize.cs), but have to be heavily sanitized, see [Add ImageSharp benchmark #2149 (comment)]
(https://github.com/dotnet/performance/pull/2149#issuecomment-986080701).

Done.

> * The jpeg decoder benchmark `DecodeJpeg_ImageSpecific` should be replaced with `DecodeJpeg` from [Unified jpeg benchmark SixLabors/ImageSharp#1844](https://github.com/SixLabors/ImageSharp/pull/1844)

Done.

> * I would consider removing the benchmarks under the folder `Color`, it's an unfinished niche feature

Done.

> would recommend updating to the [latest V2.x version](https://github.com/SixLabors/ImageSharp/tree/release/2.1.x) before continuing

Could you confirm the exact version that I will put in `ImageSharp.Benchmarks.csproj`?

kunalspathak commented:
> > Could you confirm the exact version that I will put in `ImageSharp.Benchmarks.csproj`?
> 
> 2.1.3

I tried that and getting these errors:

```
Error	NU1102	Unable to find package SixLabors.ImageSharp with version (>= 2.1.3)
  - Found 13 version(s) in dotnet-public [ Nearest version: 1.0.2 ]
  - Found 0 version(s) in benchmark-dotnet-prerelease
  - Found 0 version(s) in dotnet-eng
  - Found 0 version(s) in dotnet-tools
  - Found 0 version(s) in dotnet3.1-transport
  - Found 0 version(s) in dotnet3.1
  - Found 0 version(s) in dotnet5-transport
  - Found 0 version(s) in dotnet5
  - Found 0 version(s) in dotnet6
  - Found 0 version(s) in dotnet7	ImageSharp.Benchmarks	D:\git\performance\src\benchmarks\real-world\ImageSharp\ImageSharp.Benchmarks.csproj	1	

```

Also I am getting these errors for

```c#
this.Configuration.WorkingBufferSizeHintInBytes = this.WorkingBufferSizeHintInKilobytes * 1024;
```

```
Error	CS0281	Friend access was granted by 'SixLabors.ImageSharp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d998eea7b14cab13', but the public key of the output assembly ('') does not match that specified by the InternalsVisibleTo attribute in the granting assembly.	ImageSharp.Benchmarks (net5.0)	D:\git\performance\src\benchmarks\real-world\ImageSharp\Processing\Resize.cs	79	N/A
````

kunalspathak commented:
@antonfirsov, @JimBobSquarePants - This is ready for the review. Let me know how I can upgrade to the latest ImageSharp nuget version.

kunalspathak commented:
ping @antonfirsov, @JimBobSquarePants

JimBobSquarePants commented:
Apologies, I’ll review over the weekend

kunalspathak commented:
> Apologies, I’ll review over the weekend

I can understand things can be busy so just sending a friendly remainder in case it was forgotten. :)

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Processing/Resize.cs:
```diff
@@ -0,0 +1,129 @@
+    /// Is it worth to set a larger working buffer limit for resize?
+    /// Conclusion: It doesn't really have an effect.
+    /// </summary>
+    public class Resize_Bicubic_Rgba32_CompareWorkBufferSizes : Resize_Bicubic_Rgba32
```
This benchmark is the cause of CS0281, it's not needed for non-regression purposes.

antonfirsov commented on src/benchmarks/real-world/ImageSharp/Processing/Resize.cs:
```diff
@@ -0,0 +1,129 @@
+            => ctx.Resize(this.DestSize, this.DestSize, KnownResamplers.Bicubic, true);
+    }
+
+    public class Resize_Bicubic_Compare_Rgba32_Rgb24
```
In general, if a benchmark is about comparing A vs B, it's safe to trim it.

antonfirsov replied:
@kunalspathak this should be trimmed away,

antonfirsov reviewed (commented):
> Error	NU1102	Unable to find package SixLabors.ImageSharp with version (>= 2.1.3)

I don't see the cause of this issue, I would have to clone the repository and try it myself to figure out. Any chance you can try to investigate? I will be on a vacation the rest of the week.

kunalspathak commented:
I think I got this working with just 1 problem remaining. For some reason, I get compilation error while trying to invoke `Decode()` at https://github.com/kunalspathak/performance/blob/dae25ad21d40e7cc3781d45f9870a39ace45e86b/src/benchmarks/real-world/ImageSharp/Codecs/Jpeg/DecodeJpeg.cs#L27. It resolves to `ImageSharp 2.0` version code that navigates to this definition:

![image](https://user-images.githubusercontent.com/12488060/198646619-78cb4a46-9e89-4f00-b581-b1607debd2c9.png)

I tried clearing my nuget packages, but it doesn't help.

antonfirsov commented:
The code you took is probably from the time before we changed `IImageDecoder` to take a `CancellationToken`. What if you change it like this?
```diff
- using Image img = this.decoder.Decode(Configuration.Default, this.preloadedImageStream);
+ using Image img = this.decoder.Decode(Configuration.Default, this.preloadedImageStream, default);
```

kunalspathak commented:
> What if you change it like this?

that works. Wanted to make sure if `default` is fine with benchmark code. Any other comments or should it be good to go?

kunalspathak commented:
Ping @antonfirsov  and @JimBobSquarePants

JimBobSquarePants commented:
I cannot see anything out of the ordinary 👍

kunalspathak commented:
> I cannot see anything out of the ordinary 👍

Thanks for confirming @JimBobSquarePants and thank you for the valuable feedback. @sblom - This one is another "real-world" benchmark we would like to merge.

antonfirsov reviewed (approved):
Good in general, but noticed one point from previous review that was not addressed. Might worth a final self-review if there's anything left that can be trimmed away.

kunalspathak commented:
> one point from previous review that was not addressed. 
Fixed.

> Might worth a final self-review if there's anything left that can be trimmed away.

We are now with following benchmarks:

```
SixLabors
 └─ImageSharp
    └─Benchmarks
       ├─DetectEdges
       │  └─ImageProcessorCoreDetectEdges
       ├─Resize_Bicubic_Rgb24
       │  └─ImageSharp_P1
       ├─Resize_Bicubic_Rgba32
       │  └─ImageSharp_P1
       ├─Codecs
       │  ├─DecodeBmp
       │  │  └─BmpImageSharp
       │  ├─DecodeFilteredPng
       │  │  ├─PngFilter0
       │  │  ├─PngFilter1
       │  │  ├─PngFilter2
       │  │  ├─PngFilter3
       │  │  └─PngFilter4
       │  ├─DecodeGif
       │  │  └─GifImageSharp
       │  ├─DecodePng
       │  │  └─PngImageSharp
       │  ├─DecodeTga
       │  │  └─TgaImageSharp
       │  ├─EncodeBmp
       │  │  └─BmpImageSharp
       │  ├─EncodeGif
       │  │  └─GifImageSharp
       │  ├─EncodeIndexedPng
       │  │  ├─PngCoreOctree
       │  │  ├─PngCoreOctreeNoDither
       │  │  ├─PngCorePalette
       │  │  ├─PngCorePaletteNoDither
       │  │  ├─PngCoreWu
       │  │  └─PngCoreWuNoDither
       │  ├─EncodePng
       │  │  └─PngCore
       │  ├─EncodeTga
       │  │  └─ImageSharpTga
       │  └─Jpeg
       │     └─DecodeJpeg
       │        ├─JpegBaselineInterleaved444
       │        ├─JpegBaselineInterleaved420
       │        ├─JpegBaseline400
       │        └─JpegProgressiveNonInterleaved420
       ├─Processing
       │  ├─BokehBlur
       │  │  └─Blur
       │  ├─Crop
       │  │  └─CropImageSharp
       │  ├─Diffuse
       │  │  ├─DoDiffuse
       │  │  └─DoDither
       │  ├─HistogramEqualization
       │  │  ├─GlobalHistogramEqualization
       │  │  └─AdaptiveHistogramEqualization
       │  └─Rotate
       │     └─DoRotate
       └─Samplers
          └─GaussianBlur
             └─Blur
```
